File: 2022/day12.cpp
// https://adventofcode.com/2022/day/12

#include "../common/common.h"
#include "../common/rust.h"

#include <algorithm>
#include <iostream>
#include <print>
#include <ranges>

using Terminator = aoc::closed_range<point>;
using Input = std::pair<aoc::char_grid<>, Terminator>;

namespace stdv = std::views;
namespace stdr = std::ranges;

fn parse(String const& filename) -> Input {
  auto [heightmap, config] =
      aoc::read_char_grid(filename, aoc::char_grid_config_input{
                                        .padding = {},
                                        .start_char = 'S',
                                        .end_char = 'E',
                                    });
  // Replace begin and end with regular heights
  let terminator = Terminator{config.start_pos.value(), config.end_pos.value()};
  heightmap.modify('a', terminator.begin.y, terminator.begin.x);
  heightmap.modify('z', terminator.end.y, terminator.end.x);
  return {std::move(heightmap), terminator};
}

fn shortest_path(aoc::char_grid<> const& heightmap, point start, point end)
    -> Option<u32> {
  let distances = aoc::shortest_distances_dijkstra(
      start,
      [&](point current) {
        return heightmap.basic_neighbor_positions(current) |
               stdv::filter([&, current_height = heightmap.at(
                                    current.y, current.x)](point pos) {
                 let neighbor = heightmap.at(pos.y, pos.x);
                 return neighbor <= (current_height + 1);
               }) |
               aoc::dijkstra_uniform_neighbors_view();
      },
      end);
  let it = distances.find(end);
  if (it == std::end(distances)) {
    return {};
  }
  return it->second;
}

fn solve_case1(Input const& input) -> u32 {
  let & [ heightmap, terminator ] = input;
  return shortest_path(heightmap, terminator.begin, terminator.end).value();
}

fn solve_case2(Input const& input) -> u32 {
  let & [ heightmap, terminator ] = input;
  return stdr::min( //
      heightmap |
      stdv::enumerate |
      stdv::filter([](let& current) {
        let & [ _, value ] = current;
        return value == 'a';
      }) |
      stdv::transform([&](let& current) {
        let & [ linear_index, _ ] = current;
        return shortest_path(heightmap, heightmap.position(linear_index),
                             terminator.end)
            .value_or(heightmap.size());
      }));
}

int main() {
  std::println("Part 1");
  let example = parse("day12.example");
  AOC_EXPECT_RESULT(31, solve_case1(example));
  let input = parse("day12.input");
  AOC_EXPECT_RESULT(504, solve_case1(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(29, solve_case2(example));
  AOC_EXPECT_RESULT(500, solve_case2(input));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2022/src/day12.rs
use aoc::dijkstra::DijkstraNeighborView;
use aoc::grid::ConfigInput;
use aoc::grid::Grid;
use aoc::grid::Upos;
use aoc::iter::CollectArrayVec;
use std::ops::Range;

type Terminator = Range<Upos>;
type Input = (Grid<u8>, Terminator);

fn parse(filename: &str) -> Input {
    let (mut heightmap, config) = Grid::from_file_config(
        filename,
        ConfigInput {
            padding: None,
            start_char: Some('S'),
            end_char: Some('E'),
        },
    );
    // Replace begin and end with regular heights
    let terminator = Terminator {
        start: config.start_pos.unwrap(),
        end: config.end_pos.unwrap(),
    };
    heightmap.modify('a', terminator.start.y, terminator.start.x);
    heightmap.modify('z', terminator.end.y, terminator.end.x);
    return (
        Grid::from_iter(
            heightmap.data.iter().map(|&c| c as u8 - 'a' as u8),
            heightmap.num_rows,
            heightmap.num_columns,
        ),
        terminator,
    );
}

fn shortest_path(heightmap: &Grid<u8>, start: Upos, end: Upos) -> Option<u32> {
    let distances = aoc::dijkstra::shortest_distances(&start, &end, |current: &Upos| {
        let current_height = *heightmap.get(current.y, current.x);
        heightmap
            .basic_neighbor_positions(*current)
            .iter()
            .filter(|pos| {
                let neighbor = *heightmap.get(pos.y, pos.x);
                return neighbor <= (current_height + 1);
            })
            .cloned()
            .dijkstra_uniform_neighbors()
            .collect_array_vec::<4>()
    });
    return distances.get(&end).cloned();
}

fn solve_case1((heightmap, terminator): &Input) -> u32 {
    shortest_path(heightmap, terminator.start, terminator.end).unwrap()
}

fn solve_case2((heightmap, terminator): &Input) -> u32 {
    heightmap
        .data
        .iter()
        .enumerate()
        .filter(|(_, value)| **value == 0)
        .map(|(linear_index, _)| {
            shortest_path(heightmap, heightmap.position(linear_index), terminator.end)
                .unwrap_or(heightmap.data.len() as u32)
        })
        .min()
        .unwrap()
}

fn main() {
    println!("Part 1");
    let example = parse("day12.example");
    assert_eq!(31, solve_case1(&example));
    let input = parse("day12.input");
    assert_eq!(504, solve_case1(&input));

    println!("Part 2");
    assert_eq!(29, solve_case2(&example));
    assert_eq!(500, solve_case2(&input));
}
